ssl/quic/quic_port.c

```diff
@@ -647,9 +659,164 @@ static int port_try_handle_stateless_reset(QUIC_PORT *port, const QUIC_URXE *e)
+     * per conn_id by copying only conn_id.id_len + 1 bytes (for the id_len
+     * field) and restoring it in the same way on validation.
+    */
+    buf = OPENSSL_malloc(1 + sizeof(OSSL_TIME)
+                         + sizeof(token->odcid.id_len) + token->odcid.id_len
+                         + sizeof(token->rscid.id_len) + token->rscid.id_len
+                         + token->remote_addr_len
+                         + QUIC_RETRY_INTEGRITY_TAG_LEN);
+    if (!buf)
+        return 0;
+
+    if (token->is_retry != 0 && token->is_retry != 1)
+        return 0;
+    memset(buf, token->is_retry, 1);
+    offset += 1;
+
+    memcpy(buf + offset, &token->timestamp, sizeof(token->timestamp));
+    offset += sizeof(token->timestamp);
+
+    memcpy(buf + offset, &token->odcid.id_len, sizeof(token->odcid.id_len));
+    offset += sizeof(token->odcid.id_len);
+    memcpy(buf + offset, &token->odcid.id, token->odcid.id_len);
+    offset += token->odcid.id_len;
+
+    memcpy(buf + offset, &token->rscid.id_len, sizeof(token->rscid.id_len));
+    offset += sizeof(token->rscid.id_len);
+    memcpy(buf + offset, &token->rscid.id, token->rscid.id_len);
+    offset += token->rscid.id_len;
+
+    memcpy(buf + offset, &token->remote_addr_len,
+           sizeof(token->remote_addr_len));
+    offset += sizeof(token->remote_addr_len);
+
+    memcpy(buf + offset, token->remote_addr, token->remote_addr_len);
+    offset += token->remote_addr_len;
+
+    *buffer = buf;
+    *buffer_len = offset + QUIC_RETRY_INTEGRITY_TAG_LEN;
+    return 1;
+}
+
+/**
+ * @brief Parses contents of a buffer into a validation token.
+ *
+ * VALIDATION_TOKEN should already be initalized. Does some basic sanity checks.
+ *
+ * @param validation_token Validation token to fill data in.
+ * @param buf              Buffer of previously marshaled validation token.
+ * @param buf_len          Length of |buf|.
+ */
+static int parse_validation_token(QUIC_VALIDATION_TOKEN *validation_token,
+                                  const unsigned char *buf, size_t buf_len)
+{
+    const unsigned char *curr = buf;
+    size_t curr_len;
+    QUIC_VALIDATION_TOKEN token = { 0 };
+
+    if (buf == NULL || validation_token == NULL)
+        goto err;
+
+    token.remote_addr = NULL;
+
+    token.is_retry = *curr;
+    if (token.is_retry != 0 && token.is_retry != 1)
+        goto err;
+    curr += 1;
+
+    memcpy(&token.timestamp, curr, sizeof(token.timestamp));
+    curr += sizeof(token.timestamp);
+
+    memcpy(&token.odcid.id_len, curr, sizeof(token.odcid.id_len));
+    curr += sizeof(token.odcid.id_len);
+    memcpy(&token.odcid.id, curr, token.odcid.id_len);
+    curr += token.odcid.id_len;
+
+    memcpy(&token.rscid.id_len, curr, sizeof(token.rscid.id_len));
+    curr += sizeof(token.rscid.id_len);
+    memcpy(&token.rscid.id, curr, token.rscid.id_len);
+    curr += token.rscid.id_len;
+
+    token.remote_addr_len = *curr;
```

Comment:
I think this is wrong. the remote_addr_len is size_t, the curr is unsigned char*. I'm afraid we need a proper typecast here to fetch size_t from the address. perhaps: `*((size_t)cur)` will do the right thing for us.